### ISSUE TITLE
Questions regarding Walrus evaluation and request for numerical data

### ISSUE THREAD

**Issue by hassony2** [PolymathicAI/walrus#5]
Hi !

Thank you for the great work on the Walrus paper and thank you for releasing the code an the pretrained checkpoint :)
 
I was wondering if you could share the following details:
- Numerical results for Figure 15 In the Appendix whichshows the VRMSE loss over time for the pretrained models. This would be extremely helpful to facilitate direct comparison with the pretrained checkpoint.

- Metric Clarification. Could you confirm the calculation for the "median VRMSE". Specifically that the VRMSE first averages over physical variables and (optionally) averages across all time horizons/steps before computing the median across test trajectories ?

- Could you confirm that results for tables 7 and 8 are are computed for **finetuned** Walrus while figure 15 is for pretrained model and table 8 for finetuned models ? (In page 9 of the paper it is written "Full loss over time plots can be seen in Appendix Figure 15 while direct numerical versions of these plots can be found in Appendix Table 8.", but it seems to me that Figure 15 is on pretrained model and Table 8 is for the finetuned one as mentioned page 31 "Table 8 shows the finetuning dataset numbers").

Thank you for the clarifications you might provide :)

All the best !

Yana

**Comment by mikemccabe210** [comment]
Hi Yana,

- Probably not in this case. It's not really feasible to plug that many numbers into a paper and we're not planning on releasing a dump of all computed metrics. What we can do is make the finetuned weights used for each of these accessible so they can be easily recomputed (or used for further computations), but that will take a little while to get done. 
- Yep. The one exception to this being one-step computations where it's treated as a sliding window so the median is over all possible one-step ahead predictions over all trajectories. 
- Splitting up this since there's a couple different answers: 
  - Table 7 is just the pretrained model, but as page 5 notes, it's also a small random subset of the validation data that's being used as a quick ablation, so it's not directly comparable to full results listing. 
  - Table 8 is finetuned - those are all from out of the training set. There are no zero shot experiments in the paper.
  - Yeah, that should say F14 which is the corresponding downstream data. Thanks for pointing that out. F15 on the other hand has both. "Walrus-PT" in the key corresponding to PT only. 

Hope that helps!

Best,
Mike

**Comment by hassony2** [comment]
Thank you very much for your answer ! This clarifies my understanding :)
One follow-up question on the median computations in Table 1 of the paper. There you report medians aggregated across time steps [1:20] and [21:60].
I am trying to reuse your scripts to get comparable results.
Currently if I run  the [eval_one_gpu](https://github.com/PolymathicAI/walrus/blob/9c8b39acfae6eccf12410fb41a81818f08bce53e/walrus/run_scripts/eval_onegpu_example_walrus.sh) script with max_rollout_steps=20 and n_steps_input=6 I get rollouts of size 25 in practice because the number or roullout steps is set to `max_rollout_steps + n_steps_input - 1` for causal models [here](https://github.com/PolymathicAI/walrus/blob/9c8b39acfae6eccf12410fb41a81818f08bce53e/walrus/trainer/training.py#L439-L440). 

I therefore wanted to validate that to get comparable results, I can run the eval script with max_rollout_steps=15 (which then is converted to 20 rollout steps since `max_rollout_steps=15 + n_steps_input=6 - 1 = 20`), and use the value computed in for rollouts which is accumulated in the T=all_median [aggregated_losses](https://github.com/PolymathicAI/walrus/blob/9c8b39acfae6eccf12410fb41a81818f08bce53e/walrus/trainer/training.py#L994) 'rollout_valid_turbulent_radiative_la
yer_2D/full_VRMSE_T=all_median' field.

**Comment by mikemccabe210** [comment]
Generally for eval metrics since there's a mix of causal and acausal models with different context lengths, if the context window goes from T_s:T_e, then all validation metrics in the paper start at T_e+1, regardless of whether the model is causal or not, so you'd still need max_rollout_steps=20.  

The other important parameter there is in the data module - `start_rollout_valid_output_at_t=17` (or 11 for 3D examples) to match what was originally the longest context length included in comparisons. In practice, that would start the rollouts with a context length of `[start_rollout_valid_output_at_t-1-n_steps_input:start_rollout_valid_output_at_t-1]` and predict `[start_rollout_valid_output_at_t:start_rollout_valid_output_at_t+min(max_rollout_steps, remaining_dataset_length]`.

**Comment by hassony2** [comment]
Thank you for your answer, I am still not 100% sure on how I should proceed to get a number which is comparable to Table 1 median [1:20] in the paper.


To get comparable results, should I:
- (a) launch the command with `++trainer.**max_rollout_steps=15` and use the output 
- (b) launch the command with `++trainer.**max_rollout_steps=20`, in which case I observe 25 rollout time steps in time_logs and apply some slicing to get back to 20 time steps for the T=all median ? 

In practice, for instance, if I run your code with max_rollout_steps=**15** and start_rollout_valid_output_at_t=17 with the following command:

```sh
python train.py   --config-path "/home/yhasson/walrus/configs"   --config-name "extended_config.yaml"   ++distribution.distribution_type=local   ++finetune=False   ++validation_mode=True   checkpoint.coalesced_checkpoint_path="/home/yhasson/walrus/checkpoints/walrus.pt"   ++data.wandb_data_name="turbulent_radiative_layer_2D"   ++data.well_base_path="/home/yhasson/data/the_well"   ++data.module_parameters._target_="walrus.data.MixedWellDataModule"   ++data.module_parameters.batch_size=1   ++data.module_parameters.well_dataset_info.turbulent_radiative_layer_2D.path="/home/yhasson/data/the_well/turbulent_radiative_layer_2D"   ++data.module_parameters.well_dataset_info.turbulent_radiative_layer_2D.include_filters="[]"   ++data.module_parameters.well_dataset_info.turbulent_radiative_layer_2D.exclude_filters="[]"   '++trainer.validation_suite=[{_target_:the_well.benchmark.metrics.VRMSE}]'   '++trainer.validation_trajectory_metrics=[]'   '++trainer.batch_aggregation_fns=[torch.mean,torch.median]'  ++model.jitter_patches=True ++data.module_parameters.**start_rollout_valid_output_at_t=17**   ++trainer.**max_rollout_steps=15**
```

I observe that time_logs contains *5 time steps* (because of the max_rollout_step = self.max_rollout_step + T_in - 1 [line](https://github.com/PolymathicAI/walrus/blob/9c8b39acfae6eccf12410fb41a81818f08bce53e/walrus/trainer/training.py#L439-L440) and because walrus is causal.

In particular, in that case I get the following logs

```python
time_logs['turbulent_radiative_layer_2D']['turbulent_radiative_layer_2D/full_VRMSE_rollout'].shape
#  torch.Size([9, 20])`
aggregated_losses['rollout_valid_turbulent_radiative_layer_2D/full_VRMSE_T=all_median']
 # tensor(0.5392, device='cuda:0')
time_logs['turbulent_radiative_layer_2D']['turbulent_radiative_layer_2D/full_VRMSE_rollout'].mean(-1).median()
  # tensor(0.5392), as expected, avg across time and median for trajectories matches the aggregated_losses result
```

If I use `trainer.max_rollout_steps=20`, then if I read my logs correctly, I have T=all which is by default aggregated across 25 timesteps in aggregated_losses, which isn't the intended behavior, right ?

**Comment by mikemccabe210** [comment]
Ah, yeah, that is allowing extra steps that it shouldn't be in validation. Good catch. We generally sync the dataset max length and the rollout length (at 200) so hadn't run into that, but you should be able to just truncate the extra steps: it looks like its correctly [dropping](https://github.com/PolymathicAI/walrus/blob/9c8b39acfae6eccf12410fb41a81818f08bce53e/walrus/trainer/training.py#L486) the "within context" predictions from the causal model, so those should just be extra predictions at the end. 
```bash
srun python train.py --config-path="/path/" --config-name="extended_config.yaml" ++distribution.distribution_type=local ++auto_resume=True ++validation_mode=True ++folder_override="/path/" "++trainer.validation_suite=[{_target_:the_well.benchmark.metrics.NRMSE},{_target_:the_well.benchmark.metrics.VRMSE},{_target_:the_well.benchmark.metrics.PearsonR}]"   "++trainer.validation_trajectory_metrics=[]"   "++trainer.batch_aggregation_fns=[torch.mean,torch.median,torch.std]" "++data.module_parameters.max_rollout_steps=200"  "++data.module_parameters.start_rollout_valid_output_at_t=17" "++trainer.max_rollout_steps=200" "++trainer.video_size_multiplier=1" "++data.well_base_path=/data_path/" 
```
is the exact invocation used for that dataset (and generally the settings used for 2D eval. for 3D, `start_rollout_valid_output_at_t=11` and generally smaller steps to avoid needing to rewrite it to push trajectories off GPU.) We sliced and averaged the time logs after to report the metrics in the table since the log-spaced slicing used for logging in training results in dataset-specific lengths and the table was already too hard to read without having each dataset use different slicing.
